### PR TITLE
Fix 500 when visiting /note with no noteid (#248)

### LINKED ISSUE #248: AttributeError: 'noteid' when visiting ``/note`` route without form data
Visiting ``/note`` as a ``GET`` instead of a ``POST`` fails with ``AttributeError: 'noteid'`` in Sentry, and returns an unhandled error message to the user.

If I am reading the Sentry report right, the error exists because we are trying to use a non-existent key in the ``form`` object.

The fix will be to check if the ``noteid`` is even in form prior to attempting to using it.

Example: https://sentry.weasyldev.com/weasyl/weasyl/issues/1476/events/44447/

### PR DESCRIPTION
**Pull request by huntertur**
Fixes #248

I took the opportunity to practice writing tests for Weasyl by adding `weasyl.test.web.test_interaction`.

### CODE REVIEW

File: weasyl/errorcode.py [hunk 1]
```diff
@@ -98,6 +98,7 @@
     'noGuests': no_guest_access,
     "noCover": "No cover exists for that submission.",
     "noImageSource": "No image exists from which to create a thumbnail.",
+    "noteidInvalid": "The specified note is invalid.",
```
Comment (charmander) [inline weasyl/errorcode.py:101]: I think a new error code for every invalid parameter is overkill. (It’s also not a 403.)

File: weasyl/forms.py [hunk 2]
```diff
@@ -6,15 +6,16 @@
 _T = TypeVar("T")
 
 
-def expect_id(s: str) -> int:
+def expect_id(s: str, error_message="Unexpected") -> int:
```
Comment (charmander) [inline weasyl/forms.py:9]: No changes to this function, please. Keep it simple. `isinstance(s, str)` shouldn’t be checked on `s: str`.

File: weasyl/test/web/test_interaction.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+from weasyl import errorcode
+from weasyl.test import db_utils
+
+
+INVALID_PATHS = (
+    '/note',
+    '/note?',
+    '/note?noteid',
+    '/note?noteid=',
+    '/note?noteid=a',
```
Comment (charmander) [inline weasyl/test/web/test_interaction.py:12]: These tests are also overkill for something that’s more about integer parsing than notes.

**Review by charmander — CHANGES_REQUESTED** [review]

File: weasyl/errorcode.py (hunk 1 above)
Comment (huntertur) [reply weasyl/errorcode.py:101]: Would you prefer a switch to the existing `noteRecordMissing` ("That note doesn't seem to be in our database."), and if so, would you want this to be a 404 or remain the default 422?

File: weasyl/forms.py (hunk 2 above)
Comment (huntertur) [reply weasyl/forms.py:9]: Will revert this.

File: weasyl/test/web/test_interaction.py (hunk 3 above)
Comment (huntertur) [reply weasyl/test/web/test_interaction.py:12]: OK if I move these to a separate `test_forms.py` that just tests `expect_id` directly, or would you want most of these gone entirely?

File: weasyl/test/web/test_interaction.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+@pytest.mark.parametrize('path', INVALID_PATHS)
+def test_get_note_invalid(app, path):
+    user = db_utils.create_user()
+    cookie = db_utils.create_session(user)
+    resp = app.get(path, headers={'Cookie': cookie}, status=404)
+    assert resp.html.find(id='error_content').p.text.strip() == errorcode.error_messages['noteRecordMissing']
+
+
+@pytest.mark.usefixtures('db')
+def test_get_note_permissions(app):
```
Comment (charmander) [inline weasyl/test/web/test_interaction.py:30]: Thanks for adding this test! This is a good test.

File: weasyl/test/web/test_interaction.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        'content': 'Content 1',
+    }, headers={'Cookie': cookie1})
+
+    app.post('/notes/compose', {
+        'recipient': 'user3',
+        'title': 'Title 2',
+        'content': 'Content 2',
+    }, headers={'Cookie': cookie2})
+
+    resp = app.get('/note?noteid=1', headers={'Cookie': cookie1})
```
Comment (charmander) [inline weasyl/test/web/test_interaction.py:50]: Can’t rely on these notes having a particular id. As far as how to find that id… exercising `/notes` might not be a bad thing?

File: weasyl/test/web/test_interaction.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
+import pytest
+
+from weasyl import errorcode
+from weasyl.test import db_utils
+
+
+INVALID_PATHS = (
+    '/note',
+    '/note?noteid=1&noteid=2',
```
Comment (charmander) [inline weasyl/test/web/test_interaction.py:9]: Again, not really a thing that’s specific to notes. Regression tests are good in general, but extrapolating from this one, we will be testing basic parameter handling on every route.

File: weasyl/controllers/interaction.py [hunk 7]
```diff
@@ -97,9 +97,12 @@ def note_(request):
     if not define.is_vouched_for(request.userid):
         raise WeasylError("vouchRequired")
 
-    form = request.web_input()
+    try:
+        noteid = forms.expect_id(request.GET.getone("noteid"))
+    except (KeyError, WeasylError) as e:
```
Comment (charmander) [inline weasyl/controllers/interaction.py:102]: Nice try – this is about is good as it can be.

**Review by charmander — CHANGES_REQUESTED** [review]
I’d still recommend just not fixing or testing this particular endpoint standalone – there’s nothing about `/note` that merits this special treatment as far as I’m aware. But if you could separate out `test_forms` and `test_get_note_permissions` into their own PRs, those would be great.

File: weasyl/test/web/test_interaction.py (hunk 6 above)
Comment (huntertur) [reply weasyl/test/web/test_interaction.py:9]: This is fair. I will remove the `INVALID_PATHS`-related tests and keep `test_get_note_permissions`.

**Comment by huntertur** [comment]
Closing this PR to split out the new tests into their own PRs.

**Pull request closed without merging**